tiledb/sm/filter/filter_pipeline.cc

```diff
@@ -366,72 +405,301 @@ Status FilterPipeline::run_forward(Tile* tile) const {
+        result_cell_slab_ranges->cend();
+
+    std::function<Status(uint64_t, bool*)> skip_fn = std::bind(
+        &FilterPipeline::skip_chunk_reversal_fixed,
+        this,
+        std::placeholders::_1,
+        &cells_processed,
+        tile->cell_size(),
+        &cs_it,
+        cs_end,
+        std::placeholders::_2);
+
+    return run_reverse_internal(tile, config, &skip_fn);
+  }
+}
 
+Status FilterPipeline::run_reverse(
+    Tile* tile,
+    Tile* tile_var,
+    const Config& config,
+    const std::forward_list<std::pair<uint64_t, uint64_t>>*
+        result_cell_slab_ranges) const {
+  assert(!tile->filtered());
+  assert(tile_var->filtered());
+
+  if (!result_cell_slab_ranges) {
+    return run_reverse_internal(tile_var, config, nullptr);
+  } else {
+    ChunkedBuffer* const chunked_buffer_off = tile->chunked_buffer();
+    void* tmp_buffer;
+    RETURN_NOT_OK(chunked_buffer_off->get_contiguous(&tmp_buffer));
+    const uint64_t* const d_off = static_cast<uint64_t*>(tmp_buffer);
+    const uint64_t d_off_len = tile->size() / sizeof(uint64_t);
+    uint64_t cells_processed = 0;
+    uint64_t cells_size_processed = 0;
+    std::forward_list<std::pair<uint64_t, uint64_t>>::const_iterator cs_it =
+        result_cell_slab_ranges->cbegin();
+    std::forward_list<std::pair<uint64_t, uint64_t>>::const_iterator cs_end =
+        result_cell_slab_ranges->cend();
+
+    std::function<Status(uint64_t, bool*)> skip_fn = std::bind(
+        &FilterPipeline::skip_chunk_reversal_var,
+        this,
+        std::placeholders::_1,
+        d_off,
+        d_off_len,
+        &cells_processed,
+        &cells_size_processed,
+        &cs_it,
+        cs_end,
+        std::placeholders::_2);
+
+    return run_reverse_internal(tile_var, config, &skip_fn);
+  }
+}
+
+Status FilterPipeline::skip_chunk_reversal_fixed(
+    const uint64_t chunk_length,
+    uint64_t* const cells_processed,
+    const uint64_t cell_size,
+    std::forward_list<std::pair<uint64_t, uint64_t>>::const_iterator* const
+        cs_it,
+    const std::forward_list<std::pair<uint64_t, uint64_t>>::const_iterator&
+        cs_end,
+    bool* const skip) const {
+  assert(cells_processed);
+  assert(cs_it);
+
+  // As an optimization, don't waste any more instructions determining if
+  // we need to skip this chunk if there aren't any filters to reverse.
+  if (filters_.empty()) {
+    *skip = false;
+    return Status::Ok();
+  }
+
+  // Define inclusive cell index bounds for the chunk under consideration.
+  const uint64_t chunk_cell_start = *cells_processed;
+  assert(chunk_length % cell_size == 0);
+  assert(chunk_length >= cell_size);
+  const uint64_t chunk_cell_end =
+      chunk_cell_start + (chunk_length / cell_size) - 1;
+  *cells_processed = *cells_processed + (chunk_length / cell_size);
+
+  return skip_chunk_reversal_common(
+      chunk_cell_start, chunk_cell_end, cs_it, cs_end, skip);
+}
+
+Status FilterPipeline::skip_chunk_reversal_var(
```

Comment:
@stavrospapadopoulos This is the routine that checks for intersections between the result cells and var-sized attributes

Reply:
@stavrospapadopoulos The bug from yesterday was in this path. I had made the assumption that a chunks do not divide var-sized attributes, which is not true. For instance, if we have a var-sized attribute of N 8-byte values, the only guarantee is that chunks will not divide a single 8-byte value. The N cells may be split among multiple chunks. The refactor is intended to handle this scenario. This should be ready for perf testing on Monday.

Reply:
Ah great catch! Makes sense, now I recall.